File: codes/rust/692.前k个高频单词.rs
/*
 * @lc app=leetcode.cn id=692 lang=rust
 *
 * [692] 前K个高频单词
 */

// @lc code=start
use std::cmp::Reverse;
use std::collections::{BinaryHeap, HashMap};
impl Solution {
    pub fn top_k_frequent(words: Vec<String>, k: i32) -> Vec<String> {
        let mut map = HashMap::new();
        let mut heap = BinaryHeap::new();
        // 创建映射
        for word in words {
            *map.entry(word).or_insert(0) += 1;
        }
        // 建立堆
        for (key, value) in map.iter() {
            heap.push((Reverse(value), key));
            if heap.len() > k as usize {
                heap.pop();
            }
        }
        // 最后再遍历堆得到
        let mut result = Vec::new();
        while !heap.is_empty() {
            let item = heap.pop().unwrap().1.to_string();
            result.push(item);
        }
        result.reverse();
        result
    }
}
// @lc code=end

File: codes/java/692.前k个高频单词.java
import java.util.PriorityQueue;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.ArrayList;
import java.util.Collections;

class Item implements Comparable<Item> {
	String word;
	int count;

	public Item(String word, int count) {
		this.word = word;
		this.count = count;
	}

	@Override
	public int compareTo(Item o) {
		if (this.count == o.count) {
			// 按字典顺序
			return o.word.compareTo(this.word);
		}
		return this.count - o.count;
	}
}

class Solution {
	public List<String> topKFrequent(String[] words, int k) {
		// 构造单词-频率
		HashMap<String, Integer> map = new HashMap<>();
		for (String word : words) {
			map.put(word, map.getOrDefault(word, 0) + 1);
		}

		// 遍历映射并插入元素到大根堆
		PriorityQueue<Item> queue = new PriorityQueue<>();
		for (Map.Entry<String, Integer> item : map.entrySet()) {
			queue.offer(new Item(item.getKey(), item.getValue()));
			if (queue.size() > k) {
				queue.poll();
			}
		}

		List<String> result = new ArrayList<>();
		for (int i = 0; i < k; i++) {
			result.add(queue.poll().word);
		}
		Collections.reverse(result);
		return result;
	}
}